JS package reorganization

pringshia opened the pull request:
Implements #501 

## Overview

#### 4 new user-facing packages added as a replacement to `annotation-toolkit`:
- `@annotated/shell`
- `@annotated/bbox`
- `@annotated/keypoint`
- `@annotated/video-player`

#### 1 new dev-facing package added:
- `annotated/dev-scripts`

#### Updated Storybook w/ new packages

An updated storybook added that refers to these new packages has been added to `docs/sb`.
`docs/storybook` currently exists during this transition phase, and will be deleted in an upcoming PR.

#### yarn 3 workspaces will now be required for package administration

#### A GitHub action has been setup to validate the monorepo configuration.

## Acceptance criteria
Successfully run the storybook project with all components being pulled in from `@annotated/*` as opposed to the monolithic `annotation-toolkit` package.

```shell
# Before this PR:
$ cd packages/annotation-toolkit
$ npm run dev
$ cd ../../docs/storybook
$ npm run start

# After this PR:
$ yarn install
$ yarn build-all
$ yarn sb
```

## Details & methodology

A subpackage can now invoke the build process from `dev-scripts` instead of re-implementing on it's own.

#### Example: @annotated/shell's `package.json`:

```js
{
  "scripts": {
     "build": "dev-scripts build", // default is production build
     "build:dev": "BUILD_END=development dev-scripts build",
  }
}
```

---

### To-Do

- [x] Lay out directory structure 
- [x] Set up yarn workspaces w/ Lerna ([guide](https://medium.com/@jsilvax/a-workflow-guide-for-lerna-with-yarn-workspaces-60f97481149d))
  - [x] Can we get it working with React Storybook? https://github.com/storybookjs/storybook/issues/9527
      - **UPDATE** - got it working, albeit with **_many_** bumps along the way. Will post an issue up highlighting these with the storybook team later. 
      - ~Is this critical? Should this be timeboxed?~
      - ~If we cannot is there a mitigation strategy to adapt non-workspace projects to work with workspace projects?~
      - ~Barring all else, do we rollback and use just lerna instead, giving up on yarn workspace?~
   - [x] Migrate existing storybook configuration to new yarn berry-enabled setup
- [x] Consolidate build scripts and common logic into `@annotated/dev-scripts`. Inspo: [`kcd-scripts`](https://github.com/kentcdodds/kcd-scripts)
- [x] Set up a GitHub Action to test monorepo setup. (#574)


### Future Work
- Test publish. Type up a release guide on the publishing/deployment workflow (#499)
- Add READMEs and documentation for each subdirectory in @annotated/

JackUrb commented on packages/annotated/keypoint/src/Keypoints.jsx:
```diff
@@ -0,0 +1,308 @@
+    [hasHead, hasLeftArm, hasRightArm, hasLeftFoot, hasRightFoot]
+  );
+
+  // overwrite default joints with zuck joints
+  //   if (USE_DEFAULT_ZUCK_POSE) {
+  //     basePose.joints = {
+  //       head: [-0.015151105608258929, 0.4749763488769531],
+  //       l_shoulder: [0.22852434430803573, 1.474997329711914],
+  //       l_elbow: [0.3864998953683036, 2.617654037475586],
+  //       l_wrist: [1.03096923828125, 3.203685760498047],
+  //       r_shoulder: [-0.8992165701729911, 1.4161243438720703],
+  //       r_elbow: [-0.6856236049107143, 2.711177444458008],
+  //       r_wrist: [0.34967171805245534, 3.6373317718505858],
+  //       l_hip: [0.15915440150669644, 3.3800888061523438],
+  //       l_knee: [-0.33939078194754463, 4.758314514160157],
+  //       l_ankle: [-0.6899422781808036, 6.708499145507813],
+  //       l_toetip: [0.3141204833984375, 6.7678367614746096],
+  //       r_hip: [-0.8371791294642857, 3.3420761108398436],
+  //       r_knee: [-0.05901445661272321, 4.823992919921875],
+  //       r_ankle: [-0.38572256905691965, 6.677095794677735],
+  //       r_toetip: [0.4934029715401786, 6.592479705810547],
+  //     };
+  //   }
```
Still necessary?

JackUrb commented on packages/annotated/dev-scripts/README.md:
```diff
@@ -0,0 +1 @@
+# @annotated/dev-scripts
```
Also TODO? Overall I think that the structure of this lends well to us having strong component-based documentation, but I don't know how much of that you'd prefer to offload into the storybook (in which case perhaps we can do some kind of generator for these readmes, or via docusaurus?)

pringshia replied:
yeah I think we need to figure out where to go from here. I do like how tools like radix and blueprintjs do their documentation, e.g. https://www.radix-ui.com/docs/primitives/components/switch

[Docusaurus 2.0 has MDX integration](https://docusaurus.io/docs/next/markdown-features/react) with could allow us to write markdown docs with inline components. I'm leaning towards this option.

Definitely need to have a doc day soon lol

JackUrb commented on packages/annotated/bbox/src/RRRR/index.js:
```diff
@@ -0,0 +1,140 @@
+import { centerToTL, tLToCenter, getNewStyle, degToRadian } from "./utils";
+
+export default class ResizableRect extends Component {
+  //   static propTypes = {
+  //     left: PropTypes.number.isRequired,
+  //     top: PropTypes.number.isRequired,
+  //     width: PropTypes.number.isRequired,
+  //     height: PropTypes.number.isRequired,
+  //     rotatable: PropTypes.bool,
+  //     rotateAngle: PropTypes.number,
+  //     parentRotateAngle: PropTypes.number,
+  //     zoomable: PropTypes.string,
+  //     minWidth: PropTypes.number,
+  //     minHeight: PropTypes.number,
+  //     aspectRatio: PropTypes.oneOfType([
+  //       PropTypes.number,
+  //       PropTypes.bool
+  //     ]),
+  //     onRotateStart: PropTypes.func,
+  //     onRotate: PropTypes.func,
+  //     onRotateEnd: PropTypes.func,
+  //     onResizeStart: PropTypes.func,
+  //     onResize: PropTypes.func,
+  //     onResizeEnd: PropTypes.func,
+  //     onDragStart: PropTypes.func,
+  //     onDrag: PropTypes.func,
+  //     onDragEnd: PropTypes.func
+  //   }
```
Are these no longer necessary?

JackUrb reviewed (approved):
Overall I'm a huge fan of this refactor. I think this gives us a lot of room for growth moving forward, while also maintaining plenty of flexibility for people who want to use parts of `annotated` for their own projects.

Some questions inline (potentially short discussions) on documentation and possibly dead commented code.

s0mya commented on packages/annotated/keypoint/src/Keypoints.jsx:
```diff
@@ -0,0 +1,308 @@
+import React from "react";
```
Would a generic version that I derived from this code be useful in this package? 

https://github.com/fairinternal/sketch_rig/blob/main/ui/www/src/components/PoseEditor.tsx

pringshia replied:
yes.. this looks great! the only thing is that I don't have typescript support set up yet in `@annotated/dev-scripts`. I'd like to add that in, but it will most likely be a follow up PR. I'll update to your version of the PoseEditor in a follow-up PR along with that (this also makes for a great test case).

s0mya reviewed (approved):
Thanks for work Pratik! I'm looking forward to easily pulling in these components for the next project.

pringshia commented on .gitignore:
```diff
@@ -17,3 +17,13 @@ examples/**/build/*
+# https://next.yarnpkg.com/getting-started/qa#which-files-should-be-gitignored
+.yarn/*
+!.yarn/patches
+!.yarn/releases
```
Indeed, it's a conscious design decision to ensure that everyone on the team is using the same version: https://github.com/yarnpkg/yarn/issues/7741

It does make the repo size bigger, but ensures consistent deploys.

pringshia commented on .github/workflows/annotated-pkgs.yml:
```diff
@@ -0,0 +1,41 @@
+        with:
+          node-version: "14"
+
+      - name: Install yarn 3
```
I added some additional documentation at docs/development.md

No other real considerations. Old packages should continue to work with the old flow until they are upgraded over time, new packages will work out of the box. I'll also tag the repo with a git tag to denote when this transition occurred.

pringshia commented on docs/sb/README.md:
```diff
@@ -0,0 +1,70 @@
+# Getting Started with Create React App
```
Nice catch, removed this stale README and replaced with something more helpful and storybook specific.

pringshia commented on docs/sb/public/manifest.json:
```diff
@@ -0,0 +1,25 @@
+      "type": "image/png",
+      "sizes": "512x512"
+    }
+  ],
```
These were stale from create-react-app's initialization. They're used for when the app is saved on mobile. Doesn't apply to us so I've removed both the logos and the manifest file.

pringshia commented:
Thanks for the feedback guys. A second and third set of eyes is especially helpful in large PRs like these where it's easy to be distracted by the forest over the trees.

I cleaned up comments, added some development docs for Yarn, cleaned up stray initialization artifacts from create-react-app that was used for bootstrapping the storybook, and also finished the migration and replacement of the old storybook.

Also using this issue to track some of the larger documentation initiatives: #505
